training/src/main/scala/com/airbnb/aerosolve/training/LowRankLinearTrainer.scala

```diff
@@ -0,0 +1,329 @@
+    )
+  }
+
+  private def pickRandomOtherLabel(model: LowRankLinearModel,
+                                   posLabels: java.util.Map[java.lang.String, java.lang.Double],
+                                   posIdx: Int,
+                                   posMargin: Double,
+                                   rnd: Random,
+                                   dim: Int) : (Int, String, Double, Int) = {
+    // Pick a random other label. This can be a negative or a positive with a smaller margin.
+    var negIdx = rnd.nextInt(dim)
+    var negLabel = model.getLabelDictionary.get(negIdx).label
+    var negMargin : Double = if (posLabels.containsKey(negLabel)) posLabels.get(negLabel) else 0.0
+    var iter = 0
+    while ((negIdx == posIdx || negMargin > posMargin) && iter < dim) {
+      // we only want to pick a label that has smaller margin, we try at most dim times
+      // we try this for at most dim times
+      negIdx = rnd.nextInt(dim)
+      negLabel = model.getLabelDictionary.get(negIdx).label
+      negMargin = if (posLabels.containsKey(negLabel)) posLabels.get(negLabel) else 0.0
+      iter += 1
+    }
+    (negIdx, negLabel, negMargin, iter)
+  }
+
+  private def rankToLoss(k : Int, dim: Int, lossType: String): Float = {
+    lossType match {
+      case "uniform" => {
+        uniformRankLoss(k, dim)
+      }
+      case "non_uniform" => {
+        nonUniformRankLoss(k, dim)
+      }
+      case _ => {
+        1.0f
+      }
+    }
+  }
+
+  private def uniformRankLoss(k: Int, dim: Int): Float = {
+    // dim is the number of classes
+    // assume we always have dim > 1
+    k * 1.0f / (dim - 1.0f)
+  }
+
+  private def nonUniformRankLoss(k: Int, dim: Int): Float = {
+    var loss = 0.0f
+    for (i <- 1 to k) {
+      loss += 1.0f / i
+    }
+    loss
+  }
+
+  def setupModel(options : LowRankLinearTrainerOptions, pointwise : RDD[Example]) : LowRankLinearModel = {
+    val stats = TrainingUtils.getFeatureStatistics(options.minCount, pointwise)
+    val model = new LowRankLinearModel()
+    val featureWeights = model.getFeatureWeightVector
+    val labelWeights = model.getLabelWeightVector
+    val dict = model.getLabelDictionary
+    val embeddingSize = options.embeddingDimension
+    model.setEmbeddingDimension(embeddingSize)
+    var count : Int = 0
+    for (kv <- stats) {
+      val (family, feature) = kv._1
+      if (family == options.rankKey) {
+        val entry = new LabelDictionaryEntry()
+        entry.setLabel(feature)
+        entry.setCount(kv._2.count.toInt)
+        dict.add(entry)
+      } else {
+        if (!featureWeights.containsKey(family)) {
+          featureWeights.put(family, new java.util.HashMap[java.lang.String, FloatVector]())
+        }
+        val familyMap = featureWeights.get(family)
+        if (!familyMap.containsKey(feature)) {
+          count = count + 1
+          familyMap.put(feature, FloatVector.getGaussianVector(embeddingSize))
```

Comment:
Ideally this should be uniformly drawn from +/- 1/sqrt(embeddingSize)
http://stats.stackexchange.com/questions/47590/what-are-good-initial-weights-in-a-neural-network